### PR TITLE
Use comlink for extension host instead of JSON RPC 2

### PR DESCRIPTION
**Pull request by felixfbecker**

### CODE REVIEW

File: client/browser/src/extension/scripts/background.tsx [hunk 1]
```diff
@@ -366,17 +368,64 @@ function setDefaultBrowserAction(): void {
...
+    mergeMap(group => group.pipe(reduce((acc, cur) => ({ ...acc, ...cur }), {})))
+) as any
+
+// Create one extension host worker per endpoint pair
+endpointPairs.subscribe(({ proxy, expose }) => {
     const worker = createExtensionHostWorker()
-    worker.addEventListener('message', m => {
-        port.postMessage(m.data)
+    const clientAPIChannel = new MessageChannel()
+    const extensionHostAPIChannel = new MessageChannel()
```
Comment (felixfbecker) [inline client/browser/src/extension/scripts/background.tsx]: Do I understand this correctly?

```mermaid
graph LR
content_script[Content Script] ---|Chrome Port| background_script
background_script[Background Script] ---|MessageChannel| extension_host[Extension Host Worker]
```

If this correct, why do we need to wrap the Ports in the extension host worker and not just the Chrome port here between content page and background page?

File: client/browser/src/platform/extensionHost.ts [hunk 2]
```diff
@@ -1,103 +1,61 @@
...
-            callback(this.pending.pop()!)
-        }
-    }
-
-    public unsubscribe(): void {
-        super.unsubscribe()
-        this.callback = null
-        this.port.disconnect()
-    }
+            port.onMessage.addListener(listener as any)
```
Comment (felixfbecker) [inline client/browser/src/platform/extensionHost.ts]: Looks like this cast is incorrect - the `onMessage` listener seems to expect the port as the second argument and the message data as the first argument instead of the event object. Can we avoid the cast?

File: client/browser/src/platform/extensionHost.ts [hunk 3]
```diff
@@ -1,103 +1,61 @@
...
-                    this.pending.push(message)
-                }
-            } catch (err) {
-                this.fireError(err)
+export function endpointFromPort(name: string): Endpoint {
+    const port = chrome.runtime.connect({ name })
+    return MessageChannelAdapter.wrap({
+        send: data => port.postMessage(data),
+        addEventListener: (event, listener) => {
+            if (event !== 'message') {
```
Comment (felixfbecker) [inline client/browser/src/platform/extensionHost.ts]: I would remove this since the typings ensure that only `message` must be supported (and if not, the behaviour of EventListener is to allow you add any event listener that will simply never get called)

File: shared/src/api/extension/main.worker.ts [hunk 4]
```diff
@@ -1,26 +1,72 @@
...
+    /**
+     * Whether the endpoints should be wrapped with a {@link MessageChannelAdapter}.
+     *
+     * This is true when the messages passed on the endpoints are forwarded to/from
+     * other wrapped endpoints, like in the browser extension
+     */
+    wrapEndpoints: boolean
+}
+
+const isInitMessage = (m: any): m is InitMessage => m.endpoints && isEndpointPair(m.endpoints)
```
Comment (felixfbecker) [inline shared/src/api/extension/main.worker.ts]:
```suggestion
const isInitMessage = (value: any): value is InitMessage => value && isEndpointPair(value.endpoints)
```

File: shared/src/api/extension/main.worker.ts [hunk 5]
```diff
@@ -1,26 +1,72 @@
...
+
+const wrapMessagePort = (port: MessagePort): Endpoint =>
+    MessageChannelAdapter.wrap({
+        send: data => port.postMessage(data),
+        addEventListener: (event, listener) => port.addEventListener(event, listener),
+        removeEventListener: (event, listener) => port.removeEventListener(event, listener),
+    })
+
+const wrappedEndpoints = ({ proxy, expose }: InitMessage['endpoints']): EndpointPair => {
```
Comment (felixfbecker) [inline shared/src/api/extension/main.worker.ts]:
```suggestion
const wrapEndpoints = ({ proxy, expose }: InitMessage['endpoints']): EndpointPair => {
```

File: client/browser/src/platform/extensionHost.ts (hunk 2 above)
Comment (lguychard) [reply client/browser/src/platform/extensionHost.ts]: There's a few typing issues here, notably this one, introduced by [this commit](https://github.com/sourcegraph/comlink/commit/d974e5fb53acc51dffb927dc343df2ab4256e691) :`The 'this' context of type 'void' is not assignable to method's 'this' of type 'MessagePort'. `

Comment (felixfbecker) [reply client/browser/src/platform/extensionHost.ts]: Oh yeah, that `this` type should not be in the signature.

File: client/browser/src/extension/scripts/background.tsx (hunk 1 above)
Comment (lguychard) [reply client/browser/src/extension/scripts/background.tsx]: We could do it, if we also export `findMessageChannels` from comlink:

```typescript
// in background.tsx
const proxyWrapped = endpointFromPort(proxy) as MessagePort // the return value of MessageChannelAdapter.wrap() is Endpoint, not MessagePort, but we need to start it manually
proxyWrapped.start()
proxyWrapped.addEventListener('message', ({ data }) => {
    extensionHostAPIChannel.port1.postMessage(message.data, findMessageChannels(message.data))
})
```

This is necessary because the port returned by `MessageChannelAdapter.wrap()` will emit events where some leafs of the data may be `MessagePort` themselves, which need to be explicitly transferred when calling `postMessage()`. If the wrapping is done in the host worker, comlink hides the message port replacement logic.

This would also add one more full recursive traversal of every event's `data`.

Comment (felixfbecker) [reply client/browser/src/extension/scripts/background.tsx]: Got it, looks easier to just keep it as-is

File: client/browser/src/platform/extensionHost.ts (hunk 2 above)
Comment (felixfbecker) [reply client/browser/src/platform/extensionHost.ts]: Fixed upstream and published new package version

Comment (felixfbecker) [reply client/browser/src/platform/extensionHost.ts]: Shouldn't this still be this though?

```ts
port.onMessage.addListener(data => listener(new MessageEvent('message', { data })))
```

Comment (lguychard) [reply client/browser/src/platform/extensionHost.ts]: Yup, this works, but the following was also needed: 4fcfa6d

Comment (felixfbecker) [reply client/browser/src/platform/extensionHost.ts]: Ah, that explains why this worked for you before.

**Comment by felixfbecker** [comment]
I am sometimes seeing these errors:

image

Not sure if related but maybe a race condition in extension state sync

**Comment by felixfbecker** [comment]
@sqs @chrismwendt this is ready for review

**Comment by chrismwendt** [comment]
@felixfbecker How would you like me to review this code? Any particular pieces you think I would be particularly suited to provide insight for?

**Comment by felixfbecker** [comment]
All the pieces.
We will all work on this codebase so it's important that we all understand it and feel ownership for it. You and Quinn built this codebase or have at least worked in it more than I have. There could be bugs here I made during refactoring, simple mistakes, wrong assumptions or things I didn't think about.

File: shared/src/api/integration-test/codeEditor.test.ts [hunk 6]
```diff
@@ -62,7 +62,7 @@ describe('CodeEditor (integration)', () => {
                     },
                 },
             ])
-            await extensionHost.internal.sync()
+            await extensionAPI.internal.sync()
```
Comment (lguychard) [inline shared/src/api/integration-test/codeEditor.test.ts]: Aren't all of these calls effectively useless?

File: shared/src/api/client/api/documents.ts [hunk 7]
```diff
@@ -1,26 +1,22 @@
...
-    constructor(connection: Connection, modelTextDocuments: Observable<TextDocumentItem[] | null>) {
-        this.proxy = createProxyAndHandleRequests('documents', connection, this)
 
+    constructor(
+        private proxy: ProxyResult<ExtDocumentsAPI>,
+        modelTextDocuments: Observable<TextDocumentItem[] | null>
+    ) {
         this.subscriptions.add(
             modelTextDocuments.subscribe(docs => {
+                // tslint:disable-next-line: no-floating-promises
```
[the remark this answers is not in the document]
Comment (felixfbecker) [reply shared/src/api/client/api/documents.ts]: Any method that is proxied will return a Promise on the call site, even if it returns `void` on the definition side, because the call across threads is async. This was already true before this PR, but before we specified our interfaces manually instead of using a mapped type like `ProxyResult<>`. `$acceptDocumentData()` before looked like it just returned `void` even though on the call site it returned `Promise<void>`. Now with `ProxyResult<>` it is map-typed to return a Promise, so the linter complains here. It's not ideal that we ignore the Promise, but we can't react to it here in any way other than logging anyway, which happens automatically.

File: shared/src/api/integration-test/codeEditor.test.ts (hunk 6 above)
Comment (felixfbecker) [reply shared/src/api/integration-test/codeEditor.test.ts]: They are just as useful as they were before, and I think they were somewhat useless before. At least they are  and never were a guarantee that the messages were actually _handled_.

See discussion I tried to start here: https://sourcegraph.slack.com/archives/CCLF4R6EM/p1550494606005000

**Comment by felixfbecker** [comment]
Merging to unblock. @chrismwendt @sqs please review post-merge.

File: client/browser/src/extension/scripts/background.tsx [hunk 8]
```diff
@@ -366,17 +368,74 @@ function setDefaultBrowserAction(): void {
...
+                return {
+                    proxy: proxyPort,
+                    expose: exposePort,
+                }
+            })
+        )
+    )
+)
+
+// Create one extension host worker per endpoint pair
```
Comment (chrismwendt) [inline client/browser/src/extension/scripts/background.tsx]: Here's my understanding of how workers are spawned:

- The content script calls `createExtensionHost()`, which creates two new `chrome.runtime.Port`s: `proxy-<UUID>` and `expose-<UUID>`
- The background script was listening with `chrome.runtime.onConnect` and notices that two new `Port`s were created that match the regex `(port|expose)-`, so it spawns an extension host in a web worker and sets up communication between those `Port`s and the web worker using Comlink.

It would probably be helpful for the next person if this were expanded-on and documented somewhere (maybe a code comment in this file).

Comment (lguychard) [reply client/browser/src/extension/scripts/background.tsx]: Good point. PTAL 905d225 where I improved this.

File: shared/src/api/extension/main.worker.ts [hunk 9]
```diff
@@ -1,26 +1,64 @@
...
+        proxy: MessagePort
+        expose: MessagePort
+    }
+    /**
+     * Whether the endpoints should be wrapped with a comlink {@link MessageChannelAdapter}.
+     *
+     * This is true when the messages passed on the endpoints are forwarded to/from
+     * other wrapped endpoints, like in the browser extension.
+     */
+    wrapEndpoints: boolean
```
[the remark this answers is not in the document]
Comment (lguychard) [reply shared/src/api/extension/main.worker.ts]: Endpoints should be wrapped with `MessageChannelAdapter` when they do not support transferring `MessagePort` objects. This is the case when using `chrome.runtime.Port` objects as a transport.

I made this clearer in my doc improvements in 905d225. For more details, see https://github.com/GoogleChromeLabs/comlink/blob/master/messagechanneladapter.md

File: client/browser/src/extension/scripts/background.tsx [hunk 10]
```diff
@@ -366,17 +368,74 @@ function setDefaultBrowserAction(): void {
...
+    chrome.runtime.Port
+>(
+    handler => chrome.runtime.onConnect.addListener(handler),
+    handler => chrome.runtime.onConnect.removeListener(handler)
+).pipe(
+    groupBy(
+        port => (port.name || 'other').replace(ENDPOINT_KIND_REGEX, ''),
+        port => port,
+        group => group.pipe(bufferCount(2))
+    ),
```
[the remark this answers is not in the document]
Comment (lguychard) [reply client/browser/src/extension/scripts/background.tsx]: There is nothing linking the two ports together besides the regex above. I will agree with your assessment:
> fragile and implicit, but not so likely to break

As an alternative to a `CONTROL_PORT`, the content script could simply use `chrome.runtime.sendMessage` to send the id pair to the background script. Eg.:

```typescript
// content script
const proxyPortID = `proxy-${id}`
const exposePortID = `expose-${id}`
const proxyPort = chrome.runtime.connect({ name: proxyPortID })
const exposePort = chrome.runtime.connect({ name: exposePortID })
chrome.runtime.sendMessage({
    type: 'endpointPairCreated',
    payload: { proxyPortID, exposePortID }
}, () => {
    subscriber.next({
        proxy: endpointFromPort(proxyPort)
        expose: endpointFromPort(proxyPort)
    })
})

// background script
const PORTS = new Map<string, chrome.runtime.Port>()
chrome.runtime.onConnect.addListener(port => PORTS.set(port.name, port))

const _endpointPairs: Observable<{ proxy: chrome.runtime.Port; expose: chrome.runtime.Port }> = fromEventPattern<{
    type: string
    payload: any
}>(
    handler => chrome.runtime.onMessage.addListener(handler),
    handler => chrome.runtime.onMessage.removeListener(handler)
).pipe(
    filter(({ type }) => type === 'endpointPairCreated'),
    map(({ payload }) => {
        const proxy = PORTS.get(payload.proxyPortID)
        if (!proxy) {
            throw new Error(`No port with id ${payload.proxyPortID}`)
        }
        const expose = PORTS.get(payload.exposePortID)
        if (!expose) {
            throw new Error(`No port with id ${payload.proxyPortID}`)
        }
        return { proxy, expose }
    })
)
```

What do you think?

File: client/browser/src/platform/extensionHost.ts [hunk 11]
```diff
@@ -1,103 +1,57 @@
...
-            } catch (err) {
-                this.fireError(err)
-            }
-        })
-        port.onDisconnect.addListener(() => {
-            this.fireClose()
+        return createInPageExtensionHost({ wrapEndpoints: false })
+    }
+    const id = uuid.v4()
+    return new Observable(subscriber => {
```
[the remark this answers is not in the document]
Comment (lguychard) [reply client/browser/src/platform/extensionHost.ts]: I'd say it's worth it to keep this as an Observable, so that we can think of implementing reconnection logic if connection with the background page gets dropped.

File: shared/src/api/extension/main.worker.ts [hunk 12]
```diff
@@ -1,26 +1,64 @@
...
+}
+
+const isInitMessage = (value: any): value is InitMessage => value.endpoints && isEndpointPair(value.endpoints)
+
+const wrapMessagePort = (port: MessagePort) =>
+    MessageChannelAdapter.wrap({
+        send: data => port.postMessage(data),
+        addEventListener: (event, listener) => port.addEventListener(event, listener),
+        removeEventListener: (event, listener) => port.removeEventListener(event, listener),
+    })
```
[the remark this answers is not in the document]
Comment (lguychard) [reply shared/src/api/extension/main.worker.ts]: Comlink relies on the transferability of `MessagePort` objects which is not necessarily supported by the underlying transport (typically, this is not supported when using `chrome.runtime.Port` objects as transports).

`wrapMessagePort` assumes that the wrapped port only supports sending/receiving string data, and works around this by:
- [finding all `MessagePort` objects](https://sourcegraph.com/github.com/sourcegraph/comlink/-/blob/messagechanneladapter.ts#L42) in messages sent by comlink
- [Associating a unique ID](https://sourcegraph.com/github.com/sourcegraph/comlink/-/blob/messagechanneladapter.ts#L44-45) with each message channel, and replacing the `MessagePort` in the payload by its associated unique ID
- Sending the payload and all associated message channel IDs [as a JSON string](https://sourcegraph.com/github.com/sourcegraph/comlink/-/blob/messagechanneladapter.ts#L48) over the wrapped message port
- On the remote end, `MessagePort` objects are created for each received message channel ID, [IDs are replaced by their associated port objects](https://sourcegraph.com/github.com/sourcegraph/comlink/-/blob/messagechanneladapter.ts#L63-64) in the message payload
- Any message sent with an associated message channel ID on the wrapped port will be forwarded to its corresponding message channel

Comlink docs don't cover this and the code has zero comments, so hopefully this clarifies it a bit :)

**Comment by lguychard** [comment]
Thanks for the review @chrismwendt, I answered some of your question and opened a follow-up PR: #2484 

> Can I flip a `localStorage` flag to debug Comlink communication?

tl;dr: right now, you can't.

Messages sent over comlink are not very helpful when debugging. This is why we removed the "trace to console" option in the extension debug menu. We resolved to replace it by better, more specific logs that would be on all the time, but this has yet to be done.

> Will we be able to use the upstream comlink again at some point?

@felixfbecker has a PR open on GoogleChromeLabs/comlink: https://github.com/GoogleChromeLabs/comlink/pull/223

File: shared/src/api/extension/main.worker.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (chrismwendt) [reply shared/src/api/extension/main.worker.ts]: That makes more sense, thanks. I wonder why `MessagePort` is used rather than JSON strings. Perhaps for performance?

File: client/browser/src/extension/scripts/background.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (chrismwendt) [reply client/browser/src/extension/scripts/background.tsx]: :+1: to `sendMessage`ing the ID pair to the background script, that sounds even better.
